3.11.0 broke download of dependencies with snapshot parent

### Affected version

3.11.0

### Bug description

Old GlassFish version 5.1.0 has a documentation artifact with a snapshot parent in Maven Central. While Maven Dependency Plugin 3.10.0 was able to download and unpack it without any issues, 3.11.0 tries to resolve the parent and crashes the build, because the parent is not available anywhere.

(minor note - I use local mirror, but that is not important for the failure, we noticed it on CI first)
```
mvn clean install -f docs/publish
[INFO] Scanning for projects...
[INFO] 
[INFO] -------------< org.glassfish.main.docs:website-publisher >--------------
[INFO] Building Publish the Eclipse GlassFish Website 8.0.4-SNAPSHOT
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- clean:3.5.0:clean (default-clean) @ website-publisher ---
[INFO] Deleting /home/dmatej/work/repo/git/glassfish-new/docs/publish/target
[INFO] 
[INFO] --- enforcer:3.6.3:enforce (enforce-maven) @ website-publisher ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.version.RequireMavenVersion passed
[INFO] 
[INFO] --- dependency:3.11.0:unpack (get-website) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.11.0:unpack (get-current) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.11.0:unpack (get-8x) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.11.0:unpack (get-71x) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.11.0:unpack (get-7x) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.11.0:unpack (get-6x) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.11.0:unpack (get-5x) @ website-publisher ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.672 s
[INFO] Finished at: 2026-06-09T22:54:00+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:3.11.0:unpack (get-5x) 
on project website-publisher: 
Unable to find/resolve artifact.: 
Failed to read artifact descriptor for org.glassfish.docs:distribution:jar:5.1.0: 
The following artifacts could not be resolved: org.glassfish.main:nucleus-parent:pom:5.1.0-SNAPSHOT (absent): 
org.glassfish.main:nucleus-parent:pom:5.1.0-SNAPSHOT was not found in 
http://localhost:8888/repository/cache during a previous attempt. 
This failure was cached in the local repository and resolution is not reattempted until the update 
interval of nexus-cache has elapsed or updates are forced -> [Help 1]
```

With 3.10.0 build passed:
```
mvn clean install -f docs/publish
[INFO] Scanning for projects...
[INFO] 
[INFO] -------------< org.glassfish.main.docs:website-publisher >--------------
[INFO] Building Publish the Eclipse GlassFish Website 8.0.4-SNAPSHOT
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- clean:3.5.0:clean (default-clean) @ website-publisher ---
[INFO] Deleting /home/dmatej/work/repo/git/glassfish-new/docs/publish/target
[INFO] 
[INFO] --- enforcer:3.6.3:enforce (enforce-maven) @ website-publisher ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.version.RequireMavenVersion passed
[INFO] 
[INFO] --- dependency:3.10.0:unpack (get-website) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.10.0:unpack (get-current) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.10.0:unpack (get-8x) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.10.0:unpack (get-71x) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.10.0:unpack (get-7x) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.10.0:unpack (get-6x) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.10.0:unpack (get-5x) @ website-publisher ---
[INFO] 
[INFO] --- dependency:3.10.0:copy (hello-war) @ website-publisher ---
[INFO] 
[INFO] --- build-helper:3.6.1:add-resource (add-resource) @ website-publisher ---
[INFO] 
[INFO] --- build-helper:3.6.1:add-resource (add-legal-resource) @ website-publisher ---
[INFO] 
[INFO] --- resources:3.5.0:resources (default-resources) @ website-publisher ---
...
```

--- a/src/it/mrm/repository/broken-parent-artifact-1.0.pom
+++ b/src/it/mrm/repository/broken-parent-artifact-1.0.pom
@@ -1,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements.  See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership.  The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License.  You may obtain a copy of the License at
+  ~
+  ~   http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.apache.maven.its.dependency</groupId>
+    <artifactId>missing-parent</artifactId>
+    <version>1.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>broken-parent-artifact</artifactId>
+  <version>1.0</version>
+  <packaging>pom</packaging>
+</project>

--- a/src/it/projects/copy-broken-parent/invoker.properties
+++ b/src/it/projects/copy-broken-parent/invoker.properties
@@ -1,0 +1,18 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+invoker.goals = clean process-sources

--- a/src/it/projects/copy-broken-parent/pom.xml
+++ b/src/it/projects/copy-broken-parent/pom.xml
@@ -1,0 +1,55 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements.  See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership.  The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License.  You may obtain a copy of the License at
+  ~
+  ~   http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>org.apache.maven.its.dependency</groupId>
+  <artifactId>copy-broken-parent</artifactId>
+  <version>1.0-SNAPSHOT</version>
+
+  <build>
+    <plugins>
+      <plugin>
+        <artifactId>maven-dependency-plugin</artifactId>
+        <version>@project.version@</version>
+        <executions>
+          <execution>
+            <id>copy</id>
+            <goals>
+              <goal>copy</goal>
+            </goals>
+            <configuration>
+              <artifactItems>
+                <artifactItem>
+                  <groupId>org.apache.maven.its.dependency</groupId>
+                  <artifactId>broken-parent-artifact</artifactId>
+                  <version>1.0</version>
+                  <type>pom</type>
+                </artifactItem>
+              </artifactItems>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+</project>

--- a/src/it/projects/copy-broken-parent/verify.groovy
+++ b/src/it/projects/copy-broken-parent/verify.groovy
@@ -1,0 +1,20 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+assert new File(basedir, 'target/dependency/broken-parent-artifact-1.0.pom').isFile()

--- a/src/main/java/org/apache/maven/plugins/dependency/fromConfiguration/AbstractFromConfigurationMojo.java
+++ b/src/main/java/org/apache/maven/plugins/dependency/fromConfiguration/AbstractFromConfigurationMojo.java
@@ -46,7 +46,6 @@ import org.eclipse.aether.RepositorySystemSession;
 import org.eclipse.aether.artifact.DefaultArtifact;
 import org.eclipse.aether.repository.LocalRepository;
 import org.eclipse.aether.repository.LocalRepositoryManager;
-import org.eclipse.aether.resolution.ArtifactDescriptorException;
 import org.eclipse.aether.resolution.ArtifactResolutionException;
 import org.sonatype.plexus.build.incremental.BuildContext;
 
@@ -256,11 +255,11 @@ public abstract class AbstractFromConfigurationMojo extends AbstractDependencyMo
 
             RepositorySystemSession repositorySession = createSystemSessionForLocalRepo();
 
-            org.eclipse.aether.artifact.Artifact resolvedArtifact = resolverUtil.resolveArtifact(
+            org.eclipse.aether.artifact.Artifact resolvedArtifact = resolverUtil.resolveArtifactWithFallback(
                     artifact, getProject().getRemoteProjectRepositories(), repositorySession);
             return RepositoryUtils.toArtifact(resolvedArtifact);
 
-        } catch (ArtifactResolutionException | ArtifactDescriptorException e) {
+        } catch (ArtifactResolutionException e) {
             throw new MojoExecutionException("Unable to find/resolve artifact.", e);
         }
     }

--- a/src/main/java/org/apache/maven/plugins/dependency/utils/ResolverUtil.java
+++ b/src/main/java/org/apache/maven/plugins/dependency/utils/ResolverUtil.java
@@ -133,9 +133,37 @@ public class ResolverUtil {
         ArtifactDescriptorResult artifactDescriptorResult = repositorySystem.readArtifactDescriptor(
                 session, new ArtifactDescriptorRequest(artifact, repositories, null));
 
-        Artifact artifactToResolve = artifactDescriptorResult.getArtifact();
+        return resolveArtifactDirectly(artifactDescriptorResult.getArtifact(), repositories, session);
+    }
 
-        ArtifactRequest request = new ArtifactRequest(artifactToResolve, repositories, null);
+    /**
+     * Resolve a given artifact, falling back to its original coordinates when its descriptor cannot be read.
+     *
+     * @param artifact     an artifact to resolve
+     * @param repositories remote repositories list
+     * @param session      a repository system session
+     * @return resolved artifact
+     * @throws ArtifactResolutionException if the artifact could not be resolved
+     */
+    public Artifact resolveArtifactWithFallback(
+            Artifact artifact, List<RemoteRepository> repositories, RepositorySystemSession session)
+            throws ArtifactResolutionException {
+        try {
+            return resolveArtifact(artifact, repositories, session);
+        } catch (ArtifactDescriptorException descriptorException) {
+            try {
+                return resolveArtifactDirectly(artifact, repositories, session);
+            } catch (ArtifactResolutionException resolutionException) {
+                resolutionException.addSuppressed(descriptorException);
+                throw resolutionException;
+            }
+        }
+    }
+
+    private Artifact resolveArtifactDirectly(
+            Artifact artifact, List<RemoteRepository> repositories, RepositorySystemSession session)
+            throws ArtifactResolutionException {
+        ArtifactRequest request = new ArtifactRequest(artifact, repositories, null);
         ArtifactResult result = repositorySystem.resolveArtifact(session, request);
         return result.getArtifact();
     }

--- a/src/test/java/org/apache/maven/plugins/dependency/utils/ResolverUtilTest.java
+++ b/src/test/java/org/apache/maven/plugins/dependency/utils/ResolverUtilTest.java
@@ -20,13 +20,22 @@ package org.apache.maven.plugins.dependency.utils;
 
 import javax.inject.Provider;
 
+import java.util.Collections;
 import java.util.stream.Stream;
 
 import org.apache.maven.execution.MavenExecutionRequest;
 import org.apache.maven.execution.MavenSession;
+import org.eclipse.aether.RepositorySystem;
 import org.eclipse.aether.RepositorySystemSession;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.artifact.DefaultArtifact;
 import org.eclipse.aether.repository.RemoteRepository;
 import org.eclipse.aether.repository.RepositoryPolicy;
+import org.eclipse.aether.resolution.ArtifactDescriptorException;
+import org.eclipse.aether.resolution.ArtifactDescriptorRequest;
+import org.eclipse.aether.resolution.ArtifactDescriptorResult;
+import org.eclipse.aether.resolution.ArtifactRequest;
+import org.eclipse.aether.resolution.ArtifactResult;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.params.ParameterizedTest;
@@ -39,6 +48,9 @@ import org.mockito.junit.jupiter.MockitoExtension;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatCode;
 import static org.junit.jupiter.params.provider.Arguments.of;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 @ExtendWith(MockitoExtension.class)
@@ -55,6 +67,9 @@ class ResolverUtilTest {
 
     @Mock
     private Provider<MavenSession> sessionProvider;
+
+    @Mock
+    private RepositorySystem repositorySystem;
 
     @InjectMocks
     private ResolverUtil resolverUtil;
@@ -104,5 +119,23 @@ class ResolverUtilTest {
         assertThatCode(() -> resolverUtil.prepareRemoteRepository(null))
                 .isExactlyInstanceOf(NullPointerException.class)
                 .hasMessage("repository must be not null");
+    }
+
+    @Test
+    void resolveArtifactWithFallbackWhenDescriptorCannotBeRead() throws Exception {
+        Artifact artifact = new DefaultArtifact("groupId", "artifactId", null, "jar", "1.0");
+        ArtifactDescriptorRequest descriptorRequest = new ArtifactDescriptorRequest();
+        ArtifactDescriptorException descriptorException =
+                new ArtifactDescriptorException(new ArtifactDescriptorResult(descriptorRequest));
+        ArtifactResult artifactResult = new ArtifactResult(new ArtifactRequest()).setArtifact(artifact);
+
+        when(repositorySystem.readArtifactDescriptor(eq(repositorySystemSession), any(ArtifactDescriptorRequest.class)))
+                .thenThrow(descriptorException);
+        when(repositorySystem.resolveArtifact(eq(repositorySystemSession), any(ArtifactRequest.class)))
+                .thenReturn(artifactResult);
+
+        assertThat(resolverUtil.resolveArtifactWithFallback(artifact, Collections.emptyList(), repositorySystemSession))
+                .isSameAs(artifact);
+        verify(repositorySystem).resolveArtifact(eq(repositorySystemSession), any(ArtifactRequest.class));
     }
 }

--- a/src/test/java/org/apache/maven/plugins/dependency/utils/ResolverUtilTest.java
+++ b/src/test/java/org/apache/maven/plugins/dependency/utils/ResolverUtilTest.java
@@ -136,6 +136,8 @@ class ResolverUtilTest {
 
         assertThat(resolverUtil.resolveArtifactWithFallback(artifact, Collections.emptyList(), repositorySystemSession))
                 .isSameAs(artifact);
-        verify(repositorySystem).readArtifactDescriptor(eq(repositorySystemSession), any(ArtifactDescriptorRequest.class));
+        verify(repositorySystem)
+                .readArtifactDescriptor(eq(repositorySystemSession), any(ArtifactDescriptorRequest.class));
         verify(repositorySystem).resolveArtifact(eq(repositorySystemSession), any(ArtifactRequest.class));
+    }
 }

--- a/src/test/java/org/apache/maven/plugins/dependency/utils/ResolverUtilTest.java
+++ b/src/test/java/org/apache/maven/plugins/dependency/utils/ResolverUtilTest.java
@@ -136,6 +136,6 @@ class ResolverUtilTest {
 
         assertThat(resolverUtil.resolveArtifactWithFallback(artifact, Collections.emptyList(), repositorySystemSession))
                 .isSameAs(artifact);
+        verify(repositorySystem).readArtifactDescriptor(eq(repositorySystemSession), any(ArtifactDescriptorRequest.class));
         verify(repositorySystem).resolveArtifact(eq(repositorySystemSession), any(ArtifactRequest.class));
-    }
 }

